File: BirthdayBot/BackgroundServices/BirthdayRoleUpdate.vb
Imports System.Net
Imports System.Text
Imports Discord.WebSocket
Imports NodaTime

''' <summary>
''' Core automatic functionality of the bot. Manages role memberships based on birthday information,
''' and optionally sends the announcement message to appropriate guilds.
''' </summary>
Class BirthdayRoleUpdate
    Inherits BackgroundService

    Public Sub New(instance As BirthdayBot)
        MyBase.New(instance)
    End Sub

    ''' <summary>
    ''' Does processing on all available guilds at once.
    ''' </summary>
    Public Overrides Async Function OnTick() As Task
        Dim tasks As New List(Of Task)
        For Each guild In BotInstance.DiscordClient.Guilds
            Dim t = ProcessGuildAsync(guild)
            tasks.Add(t)
        Next

        Try
            Await Task.WhenAll(tasks)
        Catch ex As Exception
            Dim exs = From task In tasks
                      Where task.Exception IsNot Nothing
                      Select task.Exception
            Log($"Encountered {exs.Count} errors during bulk guild processing.")
            For Each iex In exs
                ' TODO probably not a good idea
                Log(iex.ToString())
            Next
        End Try

        ' TODO metrics for role sets, unsets, announcements - and how to do that for singles too?

        ' Running GC now. Many long-lasting items have likely been discarded by now.
        GC.Collect()
    End Function

    ''' <summary>
    ''' Does role and announcement processing for a single specified guild.
    ''' </summary>
    Public Async Function SingleUpdateFor(guild As SocketGuild) As Task
        Try
            Await ProcessGuildAsync(guild)
        Catch ex As Exception
            Log("Encountered an error during guild processing:")
            Log(ex.ToString())
        End Try

        ' TODO metrics for role sets, unsets, announcements - and I mentioned this above too
    End Function

    ''' <summary>
    ''' Main function where actual guild processing occurs.
    ''' </summary>
    Private Async Function ProcessGuildAsync(guild As SocketGuild) As Task
        ' Gather required information
        Dim tz As String
        Dim users As IEnumerable(Of GuildUserSettings)
        Dim role As SocketRole = Nothing
        Dim channel As SocketTextChannel = Nothing
        Dim announce As (String, String)
        Dim announceping As Boolean

        ' Skip processing of guild if local info has not yet been loaded
        If Not BotInstance.GuildCache.ContainsKey(guild.Id) Then Return

        ' Lock once to grab all info
        Dim gs = BotInstance.GuildCache(guild.Id)
        With gs
            tz = .TimeZone
            users = .Users
            announce = .AnnounceMessages
            announceping = .AnnouncePing

            If .AnnounceChannelId.HasValue Then channel = guild.GetTextChannel(gs.AnnounceChannelId.Value)
            If .RoleId.HasValue Then role = guild.GetRole(gs.RoleId.Value)
        End With

        ' Determine who's currently having a birthday
        Dim birthdays = GetGuildCurrentBirthdays(users, tz)
        ' Note: Don't quit here if zero people are having birthdays. Roles may still need to be removed by BirthdayApply.

        ' Set birthday roles, get list of users that had the role added
        ' But first check if we are able to do so. Letting all requests fail instead will lead to rate limiting.
        Dim roleCheck = CheckCorrectRoleSettings(guild, role)
        If Not roleCheck.Item1 Then
            SyncLock gs
                gs.OperationLog = New OperationStatus((OperationStatus.OperationType.UpdateBirthdayRoleMembership, roleCheck.Item2))
            End SyncLock
            Return
        End If

        Dim announcementList As IEnumerable(Of SocketGuildUser)
        Dim roleResult As (Integer, Integer) ' Role additions, removals
        ' Do actual role updating
        Try
            Dim updateResult = Await UpdateGuildBirthdayRoles(guild, role, birthdays)
            announcementList = updateResult.Item1
            roleResult = updateResult.Item2
        Catch ex As Discord.Net.HttpException
            SyncLock gs
                gs.OperationLog = New OperationStatus((OperationStatus.OperationType.UpdateBirthdayRoleMembership, ex.Message))
            End SyncLock
            If ex.HttpCode <> HttpStatusCode.Forbidden Then
                ' Send unusual exceptions to calling method
                Throw
            End If
            Return
        End Try

        Dim opResult1, opResult2 As (OperationStatus.OperationType, String)
        opResult1 = (OperationStatus.OperationType.UpdateBirthdayRoleMembership,
                     $"Success: Added {roleResult.Item1} member(s), Removed {roleResult.Item2} member(s) from target role.")

        If announcementList.Count <> 0 Then
            Dim announceOpResult = Await AnnounceBirthdaysAsync(announce, announceping, channel, announcementList)
            opResult2 = (OperationStatus.OperationType.SendBirthdayAnnouncementMessage, announceOpResult)
        Else
            opResult2 = (OperationStatus.OperationType.SendBirthdayAnnouncementMessage, "Announcement not considered.")
        End If

        SyncLock gs
            gs.OperationLog = New OperationStatus(opResult1, opResult2)
        End SyncLock
    End Function

    ''' <summary>
    ''' Checks if the bot may be allowed to alter roles.
    ''' </summary>
    Private Function CheckCorrectRoleSettings(guild As SocketGuild, role As SocketRole) As (Boolean, String)
        If role Is Nothing Then
            Return (False, "Failed: Designated role not found or defined.")
        End If

        If Not guild.CurrentUser.GuildPermissions.ManageRoles Then
            Return (False, "Failed: Bot does not contain Manage Roles permission.")
        End If

        ' Check potential role order conflict
        If role.Position >= guild.CurrentUser.Hierarchy Then
            Return (False, "Failed: Bot is beneath the designated role in the role hierarchy.")
        End If

        Return (True, Nothing)
    End Function

    ''' <summary>
    ''' Gets all known users from the given guild and returns a list including only those who are
    ''' currently experiencing a birthday in the respective time zone.
    ''' </summary>
    Private Function GetGuildCurrentBirthdays(guildUsers As IEnumerable(Of GuildUserSettings),
                                              defaultTzStr As String) As HashSet(Of ULong)
        Dim birthdayUsers As New HashSet(Of ULong)

        Dim defaultTz As DateTimeZone = Nothing
        If defaultTzStr IsNot Nothing Then
            defaultTz = DateTimeZoneProviders.Tzdb.GetZoneOrNull(defaultTzStr)
        End If
        defaultTz = If(defaultTz, DateTimeZoneProviders.Tzdb.GetZoneOrNull("UTC"))
        ' TODO determine defaultTz from guild's voice region

        For Each item In guildUsers
            ' Determine final time zone to use for calculation
            Dim tz As DateTimeZone = Nothing
            If item.TimeZone IsNot Nothing Then
                ' Try user-provided time zone
                tz = DateTimeZoneProviders.Tzdb.GetZoneOrNull(item.TimeZone)
            End If
            tz = If(tz, defaultTz)

            Dim targetMonth = item.BirthMonth
            Dim targetDay = item.BirthDay

            Dim checkNow = SystemClock.Instance.GetCurrentInstant().InZone(tz)
            ' Special case: If birthday is February 29 and it's not a leap year, recognize it on March 1st
            If targetMonth = 2 And targetDay = 29 And Not Date.IsLeapYear(checkNow.Year) Then
                targetMonth = 3
                targetDay = 1
            End If
            If targetMonth = checkNow.Month And targetDay = checkNow.Day Then
                birthdayUsers.Add(item.UserId)
            End If
        Next

        Return birthdayUsers
    End Function

    ''' <summary>
    ''' Sets the birthday role to all applicable users. Unsets it from all others who may have it.
    ''' </summary>
    ''' <returns>A list of users who had the birthday role applied. Use for the announcement message.</returns>
    Private Async Function UpdateGuildBirthdayRoles(g As SocketGuild,
                                              r As SocketRole,
                                              names As HashSet(Of ULong)) As Task(Of (IEnumerable(Of SocketGuildUser), (Integer, Integer)))
        ' Check members currently with the role. Figure out which users to remove it from.
        Dim roleRemoves As New List(Of SocketGuildUser)
        Dim roleKeeps As New HashSet(Of ULong)
        Dim q = 0
        For Each member In r.Members
            If Not names.Contains(member.Id) Then
                roleRemoves.Add(member)
            Else
                roleKeeps.Add(member.Id)
            End If
            q += 1
        Next

        ' TODO Can we remove during the iteration instead of after? investigate later...
        For Each user In roleRemoves
            Await user.RemoveRoleAsync(r)
        Next

        ' Apply role to members not already having it. Prepare announcement list.
        Dim newBirthdays As New List(Of SocketGuildUser)
        For Each target In names
            Dim member = g.GetUser(target)
            If member Is Nothing Then Continue For
            If roleKeeps.Contains(member.Id) Then Continue For ' already has role - do nothing
            Await member.AddRoleAsync(r)
            newBirthdays.Add(member)
        Next

        Return (newBirthdays, (newBirthdays.Count, roleRemoves.Count))
    End Function

    Public Const DefaultAnnounce = "Please wish a happy birthday to %n!"
    Public Const DefaultAnnouncePl = "Please wish a happy birthday to our esteemed members: %n"

    ''' <summary>
    ''' Makes (or attempts to make) an announcement in the specified channel that includes all users
    ''' who have just had their birthday role added.
    ''' </summary>
    Private Async Function AnnounceBirthdaysAsync(announce As (String, String),
                                                  announcePing As Boolean,
                                                  c As SocketTextChannel,
                                                  names As IEnumerable(Of SocketGuildUser)) As Task(Of String)
        If c Is Nothing Then
            Return "Announcement channel is undefined."
        End If

        Dim announceMsg As String
        If names.Count = 1 Then
            announceMsg = If(announce.Item1, If(announce.Item2, DefaultAnnounce))
        Else
            announceMsg = If(announce.Item2, If(announce.Item1, DefaultAnnouncePl))
        End If
        announceMsg = announceMsg.TrimEnd()
        If Not announceMsg.Contains("%n") Then announceMsg += " %n"

        ' Build sorted name list
        Dim namestrings As New List(Of String)
        For Each item In names
            namestrings.Add(FormatName(item, announcePing))
        Next
        namestrings.Sort(StringComparer.OrdinalIgnoreCase)

        Dim namedisplay As New StringBuilder()
        Dim first = True
        For Each item In namestrings
            If Not first Then
                namedisplay.Append(", ")
            End If
            first = False
            namedisplay.Append(item)
        Next

        Try
            Await c.SendMessageAsync(announceMsg.Replace("%n", namedisplay.ToString()))
            Return $"Successfully announced {names.Count} name(s)"
        Catch ex As Discord.Net.HttpException
            Return ex.Message
        End Try
    End Function
End Class

File: BackgroundServices/BirthdayRoleUpdate.cs
using BirthdayBot.Data;
using NodaTime;
using System.Text;

namespace BirthdayBot.BackgroundServices;
/// <summary>
/// Core automatic functionality of the bot. Manages role memberships based on birthday information,
/// and optionally sends the announcement message to appropriate guilds.
/// </summary>
class BirthdayRoleUpdate(ShardInstance instance) : BackgroundService(instance) {
    /// <summary>
    /// Processes birthday updates for all available guilds synchronously.
    /// </summary>
    public override async Task OnTick(int tickCount, CancellationToken token) {
        try {
            await ConcurrentSemaphore.WaitAsync(token).ConfigureAwait(false);
            await ProcessBirthdaysAsync(token).ConfigureAwait(false);
        } finally {
            try {
                ConcurrentSemaphore.Release();
            } catch (ObjectDisposedException) { }
        }
    }

    private async Task ProcessBirthdaysAsync(CancellationToken token) {
        // For database efficiency, fetch all pertinent 'global' database information at once before proceeding
        using var db = new BotDatabaseContext();
        var shardGuilds = Shard.DiscordClient.Guilds.Select(g => g.Id).ToHashSet();
        var presentGuildSettings = db.GuildConfigurations.Where(s => shardGuilds.Contains(s.GuildId));
        var guildChecks = presentGuildSettings.ToList().Select(s => Tuple.Create(s.GuildId, s));

        var exceptions = new List<Exception>();
        foreach (var (guildId, settings) in guildChecks) {
            var guild = Shard.DiscordClient.GetGuild(guildId);
            if (guild == null) continue; // A guild disappeared...?

            // Check task cancellation here. Processing during a single guild is never interrupted.
            if (token.IsCancellationRequested) throw new TaskCanceledException();

            // Stop if we've disconnected.
            if (Shard.DiscordClient.ConnectionState != ConnectionState.Connected) break;

            try {
                // Verify that role settings and permissions are usable
                SocketRole? role = guild.GetRole(settings.BirthdayRole ?? 0);
                if (role == null) continue; // Role not set.
                if (!guild.CurrentUser.GuildPermissions.ManageRoles || role.Position >= guild.CurrentUser.Hierarchy) {
                    // Quit this guild if insufficient role permissions.
                    continue;
                }
                if (role.IsEveryone || role.IsManaged) {
                    // Invalid role was configured. Clear the setting and quit.
                    settings.BirthdayRole = null;
                    db.Update(settings);
                    await db.SaveChangesAsync(CancellationToken.None).ConfigureAwait(false);
                    continue;
                }

                // Load up user configs and begin processing birthdays
                await db.Entry(settings).Collection(t => t.UserEntries).LoadAsync(CancellationToken.None).ConfigureAwait(false);
                var birthdays = GetGuildCurrentBirthdays(settings.UserEntries, settings.GuildTimeZone);

                // Add or remove roles as appropriate
                var announcementList = await UpdateGuildBirthdayRoles(guild, role, birthdays).ConfigureAwait(false);

                // Process birthday announcement
                if (announcementList.Any()) {
                    await AnnounceBirthdaysAsync(settings, guild, announcementList).ConfigureAwait(false);
                }
            } catch (Exception ex) {
                // Catch all exceptions per-guild but continue processing, throw at end.
                exceptions.Add(ex);
            }
        }
        if (exceptions.Count > 1) throw new AggregateException("Unhandled exceptions occurred when processing birthdays.", exceptions);
        else if (exceptions.Count == 1) throw new Exception("An unhandled exception occurred when processing a birthday.", exceptions[0]);
    }

    /// <summary>
    /// Gets all known users from the given guild and returns a list including only those who are
    /// currently experiencing a birthday in the respective time zone.
    /// </summary>
    public static HashSet<ulong> GetGuildCurrentBirthdays(IEnumerable<UserEntry> guildUsers, string? serverDefaultTzId) {
        var birthdayUsers = new HashSet<ulong>();

        foreach (var record in guildUsers) {
            // Determine final time zone to use for calculation
            DateTimeZone tz = DateTimeZoneProviders.Tzdb
                .GetZoneOrNull(record.TimeZone ?? serverDefaultTzId ?? "UTC")!;

            var checkNow = SystemClock.Instance.GetCurrentInstant().InZone(tz);
            // Special case: If user's birthday is 29-Feb and it's currently not a leap year, check against 1-Mar
            if (!DateTime.IsLeapYear(checkNow.Year) && record.BirthMonth == 2 && record.BirthDay == 29) {
                if (checkNow.Month == 3 && checkNow.Day == 1) birthdayUsers.Add(record.UserId);
            } else if (record.BirthMonth == checkNow.Month && record.BirthDay == checkNow.Day) {
                birthdayUsers.Add(record.UserId);
            }
        }

        return birthdayUsers;
    }

    /// <summary>
    /// Sets the birthday role to all applicable users. Unsets it from all others who may have it.
    /// </summary>
    /// <returns>
    /// List of users who had the birthday role applied, used to announce.
    /// </returns>
    private static async Task<IEnumerable<SocketGuildUser>> UpdateGuildBirthdayRoles(SocketGuild g, SocketRole r, HashSet<ulong> toApply) {
        var additions = new List<SocketGuildUser>();
        try {
            var removals = new List<SocketGuildUser>();
            var no_ops = new HashSet<ulong>();

            // Scan role for members no longer needing it
            foreach (var user in r.Members) {
                if (!toApply.Contains(user.Id)) removals.Add(user);
                else no_ops.Add(user.Id);
            }
            foreach (var user in removals) {
                await user.RemoveRoleAsync(r).ConfigureAwait(false);
            }

            foreach (var target in toApply) {
                if (no_ops.Contains(target)) continue;
                var user = g.GetUser(target);
                if (user == null) continue; // User existing in database but not in guild
                await user.AddRoleAsync(r).ConfigureAwait(false);
                additions.Add(user);
            }
        } catch (Discord.Net.HttpException ex)
            when (ex.DiscordCode is DiscordErrorCode.MissingPermissions or DiscordErrorCode.InsufficientPermissions) {
            // Encountered access and/or permission issues despite earlier checks. Quit the loop here, don't report error.
        }
        return additions;
    }

    public const string DefaultAnnounce = "Please wish a happy birthday to %n!";
    public const string DefaultAnnouncePl = "Please wish a happy birthday to our esteemed members: %n";

    /// <summary>
    /// Attempts to send an announcement message.
    /// </summary>
    public static async Task AnnounceBirthdaysAsync(GuildConfig settings, SocketGuild g, IEnumerable<SocketGuildUser> names) {
        var c = g.GetTextChannel(settings.AnnouncementChannel ?? 0);
        if (c == null) return;
        if (!c.Guild.CurrentUser.GetPermissions(c).SendMessages) return;

        string announceMsg;
        if (names.Count() == 1) announceMsg = settings.AnnounceMessage ?? settings.AnnounceMessagePl ?? DefaultAnnounce;
        else announceMsg = settings.AnnounceMessagePl ?? settings.AnnounceMessage ?? DefaultAnnouncePl;
        announceMsg = announceMsg.TrimEnd();
        if (!announceMsg.Contains("%n")) announceMsg += " %n";

        // Build sorted name list
        var namestrings = new List<string>();
        foreach (var item in names)
            namestrings.Add(Common.FormatName(item, settings.AnnouncePing));
        namestrings.Sort(StringComparer.OrdinalIgnoreCase);

        var namedisplay = new StringBuilder();
        foreach (var item in namestrings) {
            namedisplay.Append(", ");
            namedisplay.Append(item);
        }
        namedisplay.Remove(0, 2); // Remove initial comma and space

        await c.SendMessageAsync(announceMsg.Replace("%n", namedisplay.ToString())).ConfigureAwait(false);
    }
}
